### DISCUSSION TITLE
rustic interactive mode

### DISCUSSION THREAD

**Discussion by aawsome** [rustic-rs/rustic#1138] [Show and tell]
Hi all!

As I added and enhanced rustics interactive mode, I'd like to announce the first things here. Hope to get some feedback on it.

What are requirements for rustic interactive mode?
--
The interactive mode runs in your standard terminal (it is a TUI) and is built-in in the rustic binaries.

Where can I get a rustic binary with interactive mode?
--
The latest [nightly builds](https://github.com/rustic-rs/nightly) already contain the interactive mode.
The next rustic release will contain it too, of course.

How do I start the rustic interactive mode?
--
Currently, the only way to get interactive is by running `rustic snapshots -i`.
In future, other commands will get an `-i` option to directly jump into the interactive mode, if wanted.

Which functionalities are currently available interactively?
---

-> Snapshots view 

![grafik](https://github.com/rustic-rs/rustic/assets/37850842/5204d011-c243-4622-81cd-a3cc9feb3c3d)

![grafik](https://github.com/rustic-rs/rustic/assets/37850842/3bbff8b0-abed-4775-af19-9e9ed60189a2)

-> Editing of tags/labels/description/delete protection

![grafik](https://github.com/rustic-rs/rustic/assets/37850842/8e1a4142-ee21-4f17-8708-7bc434c5d42e)

-> Snapshot dir view (+ start restore from this view)

![grafik](https://github.com/rustic-rs/rustic/assets/37850842/e42a25ef-5b09-4cd1-bd0c-591dab80bd63)

**Comment by kapitainsky** [comment]
This is very useful and cool addition.

Any reason why there is no "delete/forget snapshot" command?

**Comment by aawsome** [comment]
Thanks for the feedback @kapitainsky 

> Any reason why there is no "delete/forget snapshot" command?

The reason is, it is just not yet implemented ;-) Are you thinking about a possibility to manually "mark" snapshots for removal and then perform the removing? This can actually be added very quickly. I'd also like to allow manual running a "forget" which would then do the marking according to the forget policy. 

But it's all just about implementing things which help users. So your feedback is very appreciated, I'll implement snapshot removal with high priority!

**Reply by kapitainsky** [reply]
This is absolutely not high priority:) rather something what would be nice to have.

**Reply by aawsome** [reply]
I was actually only talking about high priority for the interactive mode ;-)

**Reply by kapitainsky** [reply]
"r" - restore

I am trying to restore file/folder to specific target destination but nothing is restored there... Does it always restore "in place" despite option to specify it? Or maybe problem is that my "to" destination contains spaces?

I will play with it a bit more and try to identify in more details what is happening.

**Reply by kapitainsky** [reply]
When I run `rustic snapshots -i` there is significant delay (long enough to worry that something is frozen) without anything displayed on my screen. Maybe it would make sense to show some placeholder - "initialising" etc.

**Reply by kapitainsky** [reply]
when running restore I can only see progress:

```
        ┌restoring file contents...──┐
        │[00:15:59] 945.4 MiB/2.0 GiB│
        └────────────────────────────┘
```

I think it would be better to repeat on this screen all restore details (the same as  on confirmation screen).  When having few winsows opened with running restore it would be nice to see what is running where.

**Comment by aawsome** [comment]
About removing snapshots - I just implemented this (was pretty easy), see #1143.

> Does it always restore "in place" despite option to specify it? Or maybe problem is that my "to" destination contains spaces?

Actually should work as if you give the destination to the `rustic restore` command. But there could be bugs in, so if you find something not working, don't hesitate to open an issue!

> When I run rustic snapshots -i there is significant delay (long enough to worry that something is frozen) without anything displayed on my screen. Maybe it would make sense to show some placeholder - "initialising" etc.

Do you mean a delay until the "loading index.." and "loading snapshots" progress pop-ups are coming? In this case it could be a remote repository and the time to connect to it. But yes, we may have to improve the UX here, thanks for the feedback! :+1:

**Comment by kapitainsky** [comment]
I have left interactive mode running overnight and it the morning it was just filing console with some control characters... Looks like after long time some buffers overflowed:)


```
$ rustic snapshots -i
[INFO] using config ./rustic.toml
;83;35M^[[<35;82;35M^[[<35;81;35M^[[<35;80;35M^[[<35;79;36M^[[<35;78;36M^[[<35;77;36M^[[<35;76;37M^[[<35;75;37M^[[<35;74;38M^[[<35;73;38M^[[<35;72;39M^[[<35;71;39M^[[<35;71;40M^[[<35;70;40M^[[<35;70;41M^[[<35;69;41M^[[<35;69;42M^[[<35;68;43M^[[<35;68;44M^[[<35;67;44M^[[<35;83;23M^[[<35;84;23M^[[<35;84;24M^[[<35;85;25M^[[<35;85;26M^[[<35;86;27M^[[<35;86;28M^[[<35;86;29M^[[<35;86;30M^[[<35;86;31M^[[<35;87;31M^[[<35;87;32M^[[<35;87;33M^[[<35;86;34M^[[<35;83;26M^[[<35;83;27M^[[<35;83;28M^[[<35;82;29M^[[<35;82;30M^[[<35;82;31M^[[<35;81;31M^[[<35;81;32M^[[<35;81;33M^[[<35;80;34M^[[<35;103;44M^[[<35;103;43M^[[<35;104;43M^[[<35;104;42M^[[<35;105;41M^[[<35;105;40M^[[<35;106;40M^[[<35;106;39M^[[<35;106;38M^[[<35;107;38M^[[<35;107;37M^[[<35;107;36M^[[<35;108;36M^[[<35;109;35M^[[<35;110;35M^[[<35;112;34M^[[<35;113;33M^[[<35;115;33M^[[<35;117;31M^[[<35;119;30M^[[<35;121;29M^[[<35;124;27M^[[<35;127;24M^[[<35;129;23M^[[<35;131;22M^[[<35;134;20M^[[<35;136;18M^[[<35;138;16M^[[<35;140;14M^[[<35;142;12M^[[<35;144;11M^[[<35;146;
kptsky@DB-laptop:~/bin/restic/rustic_onedrive$ ;83;35M35;82;35M35;81;35M35;80;35M35;79;36M35;78;36M35;77;36M35;76;37M35;75;37M35;74;38M35;73;38M35;72;39M35;71;39M35;71;40M35;70;40M35;70;41M35;69;41M35;69;42M35;68;43M35;68;44M35;67;44M35;83;23M35;84;23M35;84;24M35;85;25M35;85;26M35;86;27M35;86;28M35;86;29M35;86;30M35;86;31M35;87;31M35;87;32M35;87;33M35;86;34M35;83;26M35;83;27M35;83;28M35;82;29M35;82;30M35;82;31M35;81;31M35;81;32M35;81;33M35;80;34M35;103;44M35;103;43M35;104;43M35;104;42M35;105;41M35;105;40M35;106;40M35;106;39M35;106;38M35;107;38M35;107;37M35;107;36M35;108;36M35;109;35M35;110;35M35;112;34M35;113;33M35;115;33M35;117;31M35;119;30M35;121;29M35;124;27M35;127;24M35;129;23M35;131;22M35;134;20M35;136;18M35;138;16M35;140;14M35;142;12M35;144;11M35;146;9M35;147;9M35;148;7M35;149;6M35;150;6M35;151;5M35;152;4M;18M35;105;18M35;103;19M35;101;20M35;96;22M35;91;24M35;85;26M35;80;28M35;74;31M35;68;34M35;62;36M35;57;39M35;55;40M35;51;42M35;48;44M35;78;8M35;78;9M35;78;10M35;77;13M35;76;14M35;75;17M35;74;18M35;71;22M35;69;25M35;67;27M35;65;30M35;63;32M35;61;34M35;59;36M35;57;39M35;55;41M35;54;42M35;53;43M35;52;44M35;51;44M35;98;43M35;97;41M35;96;40M35;95;36M35;94;35M35;94;33M35;93;31M35;92;30M35;92;29M35;.....
...

```

**Reply by kapitainsky** [reply]
this also allowed to identify where my restore attempt were restoring data ...

```
$ ll
total 72
drwxr-xr-x@ 7 kptsky  staff   224B Apr 28 22:35                                               ┌restore information─────────────────────────────────────────────────────────────────────────┐
drwxr-xr-x@ 3 kptsky  staff    96B Apr 28 21:25 "
-rw-------@ 1 kptsky  staff    10B Dec 12  2021 estore size: 2.0 GiB                                                                 │
-rw-------@ 1 kptsky  staff    10B Dec 12  2021 estore, 0 unchanged, 0 verified, 0 to modify, 0 additional                │
drwxr-xr-x@ 3 kptsky  staff    96B Apr 28 22:46 estoring from: e9477cc3:Users
drwxr-xr-x@ 3 kptsky  staff    96B Apr 28 22:46 estoring to: "
-rwxr-xr-x@ 1 kptsky  staff    99B Feb 22 10:48 get_repo_id.sh
-rw-r--r--@ 1 kptsky  staff   1.1K Mar  2 10:40 log
drwxr-xr-x@ 3 kptsky  staff    96B Apr 28 22:47 oceed (y
-rw-r--r--@ 1 kptsky  staff   2.0K Apr 28 19:20 rustic.glob
-rw-r--r--@ 1 kptsky  staff   870B Apr 28 21:11 rustic.toml
-rwxr-xr-x@ 1 kptsky  staff   2.8K Feb 13 01:55 rustic_backup.sh
-rw-------@ 1 kptsky  staff    10B Dec 12  2021 s:   33862 to restore, 1 to modify, 0 additional                                         │
-rwxr-xr-x@ 1 kptsky  staff   336B Apr 28 17:25 test.sh
```

As you can see above some "crazy" directories were created in my working directory.... some of them contain data I was trying to restore.... After cleanup it is the same folder as it was before I started:

```
$ ll
total 48
-rwxr-xr-x@ 1 kptsky  staff    99B Feb 22 10:48 get_repo_id.sh
-rw-r--r--@ 1 kptsky  staff   1.1K Mar  2 10:40 log
-rw-r--r--@ 1 kptsky  staff   2.0K Apr 28 19:20 rustic.glob
-rw-r--r--@ 1 kptsky  staff   870B Apr 28 21:11 rustic.toml
-rwxr-xr-x@ 1 kptsky  staff   2.8K Feb 13 01:55 rustic_backup.sh
-rwxr-xr-x@ 1 kptsky  staff   336B Apr 28 17:25 test.sh
```

**Reply by aawsome** [reply]
> As you can see above some "crazy" directories were created in my working directory...
Oh my god.. There seems to be a mess-up between the entered restore-destination string and the message to print out. Was this with a space in the path? I'll definitively have to look after this!

**Reply by kapitainsky** [reply]
Now testing it step by step.

1.
When nothing is specified in "to" field for restore content is restored into current directory.  I think it would be better maybe to assume that it is "in place" restore? Either way would be nice to mention in help what is default behaviour.

**Reply by kapitainsky** [reply]
> > Oh my god.. There seems to be a mess-up between the entered restore-destination string and the message to print out. Was this with a space in the path? I'll definitively have to look after this!

I think yes... and probably in quotes - "/path/to/rerstore destination"
trying to reproduce it now. It was lasting too long and I went to bed. In the morning screen was filled with control characters like I reported earlier. I had to CTRL-C it and only then discovered new content:)

**Reply by kapitainsky** [reply]
There is also possibility that it was some mistaken copy/paste to "to" field. But regardless I think it would be nice here to:

1. Show summary what/where is restored during progress phase - not only in confirmation screen
2. Validate "to" field - is it existing directory? or not?
3. If default is current dir then "to" field should show it in confirmation dialog - at the moment it is simply empty

**Comment by kapitainsky** [comment]
When `rustic` generates warning they are scattered randomly across all screen - example:

```



                                    ┌restore done─────────────────────────────────────────────────────────────────────────────────────────────────────┐
                                    │restored a6ec8d78:Users/testing_rustic_1/Temp/rustic test successfully to /Volumes/DB MacBook Pro HD Temp/restore│
                                    └─────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘                                   [WARN] Note: additional entries exist in destination


                                                                                                                                   [WARN] Note: additional entries exist in destination





```

**Comment by kapitainsky** [comment]
```
│     Del : toggle to-delete for snapshot(s)                            │                                                       █
│Ctrl-Del : clear to-delete for snapshot(s)
```

Thank you for implementing it.

Could we change the shortcut? there is no `Del` button on mac keyboards.... It can be entered by pressing Fn+Backspace but not everybody knows it...

What about f/ctrl-f and change wording from delete -> forget

**Reply by kapitainsky** [reply]
Small cosmetics - after pressing `w`:

<img width="583" alt="image" src="https://github.com/rustic-rs/rustic/assets/15635386/fc0f02dd-1636-4162-81c2-713188c8ed8c">

I assume it is `Y` I should press to proceed and indeed it works. But would be nice to state it in this dialog - e.g.: Press Y to proceed or anything else to abort.

**Comment by kapitainsky** [comment]
```

                                                                                 ┌restoring file content┐
                                                                                 │[00:00:00] 0 B/2.5 MiB│
                                                                                 └──────────────────────┘Well, this is embarrassing.

                                                                                                                                    rustic-rs had a problem and crashed. To help us diagnose the problem you can send us a crash report.

                                              We have generated a report file at "/var/folders/36/02d6s6313zq5k3mfrhg4d8zc0000gn/T/report-6d57555d-a05e-435a-828e-718c152a0bf9.toml". Submit an issue or email with the subject of "rustic-rs Crash Report" and include the report as an attachment.

                                                                                                          - Homepage: https://rustic.cli.rs/
                                                                                                                                            - Authors: the rustic-rs team

                                                                                                                                                                         We take privacy seriously, and do not perform any automated error collection. In order to improve the software, we rely on people to submit reports.

                                                                                                                                   Thank you kindly!
                                                                                                                                                    Abort trap: 6
kptsky@DB-laptop:~/bin/restic/rustic_onedrive$ 35;107;35M35;108;35M35;108;34M35;109;34M35;110;34M35;111;34M35;112;34M35;113;34M35;114;33M35;115;33M35;116;33M35;117;33M35;118;33M35;119;33M35;120;33M35;121;33M35;122;33M35;123;33M35;124;33M35;125;33M35;126;33M35;125;33M35;124;33M35;123;33M35;122;33M35;121;33M35;120;33M35;119;33M35;118;33M35;117;33M^C
kptsky@DB-laptop:~/bin/restic/rustic_onedrive$ 35;116;33M35;115;33M35;114;33M35;113;33M35;112;33M35;111;33M35;110;32M35;109;32M35;108;32M35;108;31M35;107;31M35;107;30M35;107;29M35;107;28M35;108;28M35;109;27M35;110;27M35;110;26M35;111;26M35;111;25M35;112;25M^C
kptsky@DB-laptop:~/bin/restic/rustic_onedrive$ 65;112;25M65;112;25M65;112;25M65;112;25M65;112;25M^C
kptsky@DB-laptop:~/bin/restic/rustic_onedrive$ ^C
kptsky@DB-laptop:~/bin/restic/rustic_onedrive$ 35;111;25M35;110;25M35;109;25M35;108;25M35;107;25M35;107;24M35;106;24M35;105;24M35;104;24M35;103;24M35;102;24M35;101;24M35;100;24M35;100;23M35;99;23M35;98;23M35;97;23M35;96;23M35;95;22M35;94;22M35;93;22M35;92;22M35;91;21M
```

rustic crashed during restore - it can happen. Not big deal.

But what happens in interactive mode is that now it prints some control chars forever... CTRL-C does not help. It returns to prompt but keeps printing.... I have to kill terminal window to stop it:)

**Reply by kapitainsky** [reply]
Other thing is that this panic is IMO related to interactive mode. I am trying to restore single file without providing "to" field (so should be current folder?)

Below content of generated panic report:

```
$ cat /var/folders/36/02d6s6313zq5k3mfrhg4d8zc0000gn/T/report-6d57555d-a05e-435a-828e-718c152a0bf9.toml
name = "rustic-rs"
operating_system = "Mac OS 14.4.1 [64-bit]"
crate_version = "0.7.0-dev"
explanation = """
Panic occurred in file '/Users/kptsky/.cargo/git/checkouts/rustic_core-fa42f42e094d79df/6495674/crates/core/src/commands/restore.rs' at line 539
"""
cause = "called `Result::unwrap()` on an `Err` value: ErrorSettingLength(\"WechatIMG121317.jpeg\", RusticError(LocalDestination(FileDoesNotHaveParent(\"\"))))"
method = "Panic"
backtrace = """

   0: 0x1033a2745 - __mh_execute_header
   1: 0x1033a2c65 - __mh_execute_header
   2: 0x1030a8ebc - __mh_execute_header
   3: 0x1033b6809 - __mh_execute_header
   4: 0x102b38d0c - __mh_execute_header
   5: 0x102b38abb - __mh_execute_header
   6: 0x1032289f9 - __mh_execute_header
   7: 0x7ff80625818b - __pthread_start"""
```

What I do:

```
┌restore a6ec8d78:Users/kptsky/Temp/rustic test/WechatIMG121319.jpeg to:───────────────────────────────────────────────────────────────────────────────────────────────────────┐
│enter restore destination                                                                                                                                                               │
└───────────────────────────────────────────────────────────────────────────────────
```
```

                                                  ┌restore information──────────────────────────────────────────────────────────────────┐
                                                  │                                                                                     │
                                                  │restoring from: a6ec8d78:Users/kptsky/Temp/rustic test/WechatIMG121319.jpeg.         │
                                                  │restoring to:                                                                        │
                                                  │                                                                                     │
                                                  │Files:  1 to restore, 0 unchanged, 0 verified, 0 to modify, 0 additional             │
                                                  │Dirs:   0 to restore, 0 to modify, 0 additional                                      │
                                                  │Total restore size: 3.6 MiB                                                          │
                                                  │                                                                                     │
                                                  │Do you want to proceed (y/n)?                                                        │
                                                  │                                                                                     │
                                                  └─────────────────────────────────────────────────────────────────────────────────────┘

```

Maybe it does not like space is source path?

**Comment by aawsome** [comment]
@kapitainsky Thanks again for your thorough tests and your feedback!

About the points:
- to-forget now can be controlled by f / Ctrl-f
- There is now a (y/n) at the "do you want to write..." prompt. 
- panics should now be handled better and no longer "destroy" the terminal
- Empty target is now treated as "." (this should prevents the panic in restore)
- Restore target is now pre-filled by the selected path, but can be changed by the user
- I tested restore targets with a space - worked without a problem. If you do encounter a problem, please report!

Open points:
- delay at the start -> Can you please give me the feedback if this is coming before the "reading index.." progress?
- Warning and generally all log message are printed out somewhere (messy) -> I want to overwork the log output for the interactive mode, but this is a bigger issue, so we'll have to live with the current behavior ATM.
- Print information about what is restored during the actual restore -> this point is somewhat related to the previous point. Currently, I don't have much influence about what to write in the progress as progress information is handled similar to log message. So this will also stay open a bit of time...

**Reply by kapitainsky** [reply]
Impressive pace of tackling problems! Will start using it tomorrow again and report anything I notice.

> * delay at the start -> Can you please give me the feedback if this is coming before the "reading index.." progress?

Yes - before "reading index.." progress. After executing `rustic snapshots -i` there is empty screen - nothing on it for about 10-12s - only after I see:

```
[INFO] repository rclone:OD01-ENC:restic-test: password is correct.
                                                                   [INFO] using cache at /Users/kptsky/Library/Caches/rustic/5f2d3e038efab8f03b2c1b3281e3bda9ae694ee9a03e27baba1a4f380afae80c


                                                                                     ┌reading index.┐
                                                                                     │[00:00:00] 0/0│
                                                                                     └──────────────┘

```

which shows sort of another small issue - [INFO] and similar messages IMO should be suppressed or maybe displayed in some controlled manner at the bottom of the screen? At the moment it looks like they appear randomly at random places..

This is followed by snapshots view.

> Warning and generally all log message are printed out somewhere (messy) -> I want to overwork the log output for the interactive mode, but this is a bigger issue, so we'll have to live with the current behavior ATM.

OK. This explains above [INFO] messages - I will ignore it  then.

> Print information about what is restored during the actual restore -> this point is somewhat related to the previous point. Currently, I don't have much influence about what to write in the progress as progress information is handled similar to log message. So this will also stay open a bit of time...

I was more thinking here about leaving most of the information shown on confirmation screen so progress screen would look like:

```
   ┌restore information──────────────────────────────────────────────────────────────────┐
   │                                                                                     │
   │restoring from: a6ec8d78:Users/kptsky/Temp/rustic test/WechatIMG121319.jpeg.         │
   │restoring to:                                                                        │
   │                                                                                     │
   │Files:  1 to restore, 0 unchanged, 0 verified, 0 to modify, 0 additional             │
   │Dirs:   0 to restore, 0 to modify, 0 additional                                      │
   │Total restore size: 3.6 MiB                                                          │
   │                                                                                     │
   └─────────────────────────────────────────────────────────────────────────────────────┘


                           ┌restoring file content┐
                           │[10:20:33] 0 B/3.6 MiB│
                           └──────────────────────┘

```

When I have 10 terminal windows running different restore jobs lasting hours I would love to have always on the screen info showing what is running.
And maybe instead of `restoring file content` use `restore progress` ??
But the key is information what is running here - restoring what to where.

**Reply by kapitainsky** [reply]
> About the points:
> 
>     * to-forget now can be controlled by f / Ctrl-f
> 
>     * There is now a (y/n) at the "do you want to write..." prompt.
> 
>     * panics should now be handled better and no longer "destroy" the terminal
> 
>     * Empty target is now treated as "." (this should prevents the panic in restore)
> 
>     * Restore target is now pre-filled by the selected path, but can be changed by the user
> 
>     * I tested restore targets with a space - worked without a problem. If you do encounter a problem, please report!


FYI - I have validated every single point from the above list and now it all works.

Let's see how other things stand with my usage.

**Reply by kapitainsky** [reply]
`rustic.git#a6bd54c7`

Unfortunately still after running some longish restore I ended up with this farewell screen.... no idea what causes it.

```
$ rustic snapshots -i
[INFO] using config ./rustic.toml
[<35;147;31M^[[<35;146;31M^[[<35;144;31M^[[<35;143;31M^[[<35;140;31M^[[<35;137;30M^[[<35;135;30M^[[<35;132;30M^[[<35;129;30M^[[<35;125;30M^[[<35;122;30M^[[<35;119;30M^[[<35;116;30M^[[<35;112;29M^[[<35;109;29M^[[<35;106;29M^[[<35;103;29M^[[<35;99;29M^[[<35;96;28M^[[<35;92;28M^[[<35;88;28M^[[<35;85;28M^[[<35;81;27M^[[<35;78;27M^[[<35;74;27M^[[<35;71;27M^[[<35;68;27M^[[<35;66;27M^[[<35;63;27M^[[<35;60;27M^[[<35;58;27M^[[<35;55;27M^[[<35;53;27M^[[<35;51;27M^[[<35;49;27M^[[<35;47;27M^[[<35;45;27M^[[<35;44;27M^[[<35;43;27M^[[<35;42;27M^[[<35;41;27M^[[<35;1;23M^[[<35;3;24M^[[<35;4;24M^[[<35;6;24M^[[<35;8;24M^[[<35;9;25M^[[<35;10;25M^[[<35;12;25M^[[<35;13;26M^[[<35;14;26M^[[<35;15;26M^[[<35;16;26M^[[<35;17;27M^[[<35;18;27M^[[<35;19;27M^[[<35;20;28M^[[<35;21;28M^[[<35;22;29M^[[<35;22;30M^[[<35;23;31M^[[<35;23;32M^[[<35;23;33M^[[<35;24;32M^[[<35;25;32M^[[<35;26;32M^[[<35;28;32M^[[<35;29;32M^[[<35;30;32M^[[<35;31;32M^[[<35;32;32M^[[<35;34;32M^[[<35;35;32M^[[<35;36;32M^[[<35;37;32M^[[<35;38;32M^[[<35;39;32M^[[<35;41;32M^
kptsky@DB-laptop:~/bin/restic/rustic_onedrive$ [<35;147;31M35;146;31M35;144;31M35;143;31M35;140;31M35;137;30M35;135;30M35;132;30M35;129;30M35;125;30M35;122;30M35;119;30M35;116;30M35;112;29M35;109;29M35;106;29M35;103;29M35;99;29M35;96;28M35;92;28M35;88;28M35;85;28M35;81;27M35;78;27M35;74;27M35;71;27M35;68;27M35;66;27M35;63;27M35;60;27M35;58;27M35;55;27M35;53;27M35;51;27M35;49;27M35;47;27M35;45;27M35;44;27M35;43;27M35;42;27M35;41;27M35;1;23M35;3;24M35;4;24M35;6;24M35;8;24M35;9;25M35;10;25M35;12;25M35;13;26M35;14;26M35;15;26M35;16;26M35;17;27M35;18;27M35;19;27M35;20;28M35;21;28M35;22;29M35;22;30M35;23;31M35;23;32M35;23;33M35;24;32M35;25;32M35;26;32M35;28;32M35;29;32M35;30;32M35;31;32M35;32;32M35;34;32M35;35;32M35;36;32M35;37;32M35;38;32M35;39;32M35;41;32M35;42;32M35;43;32M35;44;32M35;45;32M35;46;32M35;47;32M35;^C
```

only difference now is that CTRL-C stops printing this.

**Reply by aawsome** [reply]
@kapitainsky Thanks for reporting this issue. I have to admit that I don't have a clue what's going on there. Need to make some tests myself to see if I can reproduce that.

**Comment by kapitainsky** [comment]
Small UI improvement suggestion.

leaving interactive mode after hitting ESC - maybe it would be wise to ask for confirmation? For large repos reading index can be quite long so accidental exit (very easy now) is painful when app has to be started again.

**Reply by aawsome** [reply]
is added.

**Comment by kapitainsky** [comment]
`rustic.git#a6bd54c7`

restore

1. Looks like there is no way to terminate unwanted restore - neither CTRL-C nor ESC work

2. Guessing this is metadata restore phase -  title is truncated:

```
┌setting me┐
│[00:00:00]│
└──────────┘


```

**Reply by aawsome** [reply]
Yes, I'm aware of these points. I think those will be both tackled with the overworking of logs and progresses.

**Comment by aawsome** [comment]
@kapitainsky About

> Yes - before "reading index.." progress. After executing `rustic snapshots -i` there is empty screen - nothing on it for about 10-12s - only after I see:

is this the same behavior in non-interactive mode, i.e. does it also take 10-12s until the 
> [INFO] repository rclone:OD01-ENC:restic-test: password is correct.

is shown? Or is the 10-12s an interactive-only thing?

**Reply by kapitainsky** [reply]
yes it is the same behaviour when e.g. running `rustic snapshots`. It takes time probably to connect and "initialize" repo which in this case is rclone union of many accounts.
